### ISSUE TITLE
Best practices for dispatching actions

### ISSUE THREAD

**Issue by bzums** [ngrx/store#166]
Hey all,

tldr:
**I guess in the end my questions boils down to how to handle situations where there is one "action" (some async action, usually triggered by the user) which will update multiple parts of our state (which means multiple `reducer` will have to do something).**

I have a few general questions about how and where to dispatch actions in rather big applications. The use case is the following: 

There are actions which potentially lead to changes in multiple components of the app. Say our app has the entity `content`. There is `ContentService` which abstracts the dispatching of actions for us and `ContentApi` which abstracts http requests. 

After a http request we want to update the `content` somehow in the content reducer. But we also want to show a notification to the user that the action was successful/failed.

So we have also the entity `notification` (and a `NotificationService`). My question is now how/where to dispatch the `showNotification` action.

We could add the `NotificationService` as a dependency to `ContentService` and call some method on it at the given point in time which will dispatch an action:

``` typescript
class ContentService {
  constructor(..., private notificationService: NotificationService) {
    // ...
  }

  //...

  get(...) {
    contentApi.get(...)
     .subscribe(content => {
        this.store.dispatch(getContentSuccessAction);
        notificationService.show('success');
     })
     .catch(_ => {
        this.store.dispatch(getContentFailAction);
        notificationService.show('fail');
     })
  }
}
```

But we will end up doing that a lot of times in different services and there methods. I have the feeling that we introduce a lot of dependencies which makes it harder to test the application and also harder to  comprehend whats going on.

What we could also do is to keep the `NotificationService` out of the `ContentService` and do some additional work in the `NotificationReducer`, say we do not call the `NotificationService` explicitly but in the `NotificationReducer` we react to the `getContentSuccessAction` like that:

``` typescript
export const failureReducer = (state: NotificationState = initialState, action: Action): NotificationState => {

  switch (action.type) {
    case GET_CONTENT_SUCCESS_ACTION:
      return [...state, new Notification('Some success message')];
     // ...
  }
}
```

The problem here is that in some cases we will uses the `NotificationService` to add Notifications and some times do the way I just described which feels totally wrong. 

So after writing all this I have the feeling that the first approach is ok, but is there another option? Any hints what to improve? 

I guess in the end my questions boils down to how to handle situations where there is one "action" (some async action, usually triggered by the user) which will update multiple parts of our state (which means multiple `reducer` will have to do something).

Sorry for the long blabla, at least it was helpful to write about it but I would appreciate any feedback!

**Comment by littleStudent** [comment]
I think both options are valid to use. Using services and combining them to be able to react to different things may give you more control, but also results in a bit more complex setup.

Did you take a look at [ngrx/effects](https://github.com/ngrx/effects)? The effects library is built to handle side-effects such as http requests in a ngrx way. Maybe thats helps if you are not happy with your current solution.

**Comment by MikeRyanDev** [comment]
@bzums Re-open this issue if you are still having problems. Alternatively, feel free to ask this question in our [community Gitter channel](https://gitter.im/ngrx/store) or StackOverflow. Thanks!
